presto-base-jdbc/src/main/java/io/prestosql/plugin/jdbc/CachingJdbcClient.java

```diff
@@ -59,20 +66,31 @@
     private final Cache<TableNamesCacheKey, List<SchemaTableName>> tableNamesCache;
     private final Cache<TableHandleCacheKey, Optional<JdbcTableHandle>> tableHandleCache;
     private final Cache<ColumnsCacheKey, List<JdbcColumnHandle>> columnsCache;
+    private final List<PropertyMetadata> sessionPropertiesProviders;
 
     @Inject
-    public CachingJdbcClient(@StatsCollecting JdbcClient delegate, BaseJdbcConfig config)
+    public CachingJdbcClient(@StatsCollecting JdbcClient delegate, BaseJdbcConfig config, Set<SessionPropertiesProvider> sessionPropertiesProviders)
     {
-        this(delegate, config.getMetadataCacheTtl(), config.isCacheMissing());
+        this(delegate, sessionPropertiesProviders, config.getMetadataCacheTtl(), config.isCacheMissing());
     }
 
-    public CachingJdbcClient(JdbcClient delegate, Duration metadataCachingTtl, boolean cacheMissing)
+    public CachingJdbcClient(JdbcClient delegate, Set<SessionPropertiesProvider> sessionPropertiesProviders, Duration metadataCachingTtl, boolean cacheMissing)
     {
         this.delegate = requireNonNull(delegate, "delegate is null");
+        this.sessionPropertiesProviders = requireNonNull(sessionPropertiesProviders, "sessionPropertiesProviders is null").stream()
+            .flatMap(provider -> provider.getSessionProperties().stream())
+            .collect(toImmutableList());
+
         this.cacheMissing = cacheMissing;
 
         CacheBuilder<Object, Object> cacheBuilder = CacheBuilder.newBuilder()
                 .expireAfterWrite(metadataCachingTtl.toMillis(), TimeUnit.MILLISECONDS);
+
+        if (metadataCachingTtl.equals(succinctNanos(0L))) {
+            // Disables the cache entirely
+            cacheBuilder.maximumSize(0);
```

Comment:
I was thinking to avoid to even bind `CachingJdbcClient` in guice, so there would be no question about the way the cache is used in such case. However, I think that might be ok. All I would like to see to to write a unit test where two threads are using disabled `CachingJdbcClient` and none is waiting for the other.

Please either separate that to different commit or mention this in the commit message.

Reply:
I'm not because I've measured it and it adds the overhead of around 25 nanoseconds (since we are iterating only over catalog properties and retrieving session property is just taking it from a map and decoding (which is basically casting to a given type)

Reply:
It is not only about overhead, but also about correctness. In case there is a bug in caching we don't have a kill switch as we always do "caching". However this could go as a separate PR (or postponed until the moment we learn it is an issue).

Reply:
I've tested that cache can be disabled entirely (there is always a load)

Reply:
What about two concurrent load for the same key? Is this synchronized?

Reply:
`The concurrent hash map implementation built by CacheBuilder`. I don't see a reason to test for concurrency here (it should be covered in Guava).